kore/src/Kore/Step/Simplification/AndTerms.hs

```diff
@@ -551,6 +559,51 @@ noConfusionInjectionConstructor
 noConfusionInjectionConstructor =
     explainAndReturnBottom "No confusion: sort injections and constructors"
 
+{- |
+ If the two constructors form an overload pair, apply the overloading axioms
+ on the terms to make the constructors equal, then retry unification on them.
+
+See <https://github.com/kframework/kore/blob/master/docs/2019-08-27-Unification-modulo-overloaded-constructors.md>
+
+ -}
+overloadedConstructorSortInjectionAndEquals
+    :: (InternalVariable variable, MonadUnify unifier)
+    => TermSimplifier variable unifier
+    -> TermLike variable
+    -> TermLike variable
+    -> MaybeT unifier (Pattern variable)
+overloadedConstructorSortInjectionAndEquals termMerger firstTerm secondTerm
+  = do
+    eunifier <- lift . Error.runExceptT
+        $ unifyOverloading (Pair firstTerm secondTerm)
+    case eunifier of
+        Right (Simple (Pair firstTerm' secondTerm')) -> lift $
+            termMerger firstTerm' secondTerm'
+        Right
+            (WithNarrowing Narrowing
+                { narrowedVar
+                , narrowingTerm
+                , overloadPair = Pair firstTerm' secondTerm'
+                }
+            ) -> lift $ do
+                let narrowingVars =
+                        getFreeElementVariables $ freeVariables narrowingTerm
+                let narrowedSubst =
+                        Predicate.markSimplified
```

Comment:
Will marking this as simplified prevent it from being moved to the substitution?

Reply:
I'm not sure. @virgil-serbanuta ?

Reply:
Well... it's a bit complicated. In principle, yes, but:
1. Any change will reset the simplified bit (e.g. if we apply a substitution after computing `merged` below).
2. If we don't apply substitutions, the `Exists.makeEvaluate` call below will transform this back into a predicate (or a term).

Anyway, I discussed this with Traian and my guess is that he will make it a substitution.

Reply:
After talking with Virgil, we decided that maybe it's better if we had substitution here, as you initially proposed.